src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs

```diff
@@ -93,7 +93,14 @@ IWorkspaceProjectContext IWorkspaceProjectContextFactory.CreateProjectContext(st
             }
 
             // CPSProject constructor has a UI thread dependencies currently, so switch back to the UI thread before proceeding.
-            return new CPSProject(visualStudioProject, _workspace, _projectCodeModelFactory, projectGuid, binOutputPath);
+            var project = new CPSProject(visualStudioProject, _workspace, _projectCodeModelFactory, projectGuid);
+
+            // Set the output path in a batch, so that way the reference conversion doesn't block on the global lock
+            project.StartBatch();
+            project.BinOutputPath = binOutputPath;
+            await project.EndBatchAsync().ConfigureAwait(false);
```

Comment:
fwiw, this feels very subtle.  like why would batching make anything better, since surely it also needs to lock when making changes itself.

Reply:
maybe it's more like: this won't contend with other batched work as batched chunks run sequentially? (that's an assumption on my part btw, no clue if it's true).

Reply:
I'll revise the comment: if we didn't start a batch, the lock is acquired synchronously under the property set; with the batch it doesn't happen until we apply the batch async, so we aren't blocking everything up.

Reply:
Comment revised:

            // Set the output path in a batch; if we set the property directly we'll be taking a synchronous lock here and
            // potentially block up thread pool threads. Doing this in a batch means the global lock will be acquired 
            // asynchronously.